keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,722 @@
+    the complete list of both IPv4 and IPv6 addresses.
+
+On bare metal, `Node.Status.Addresses` will be set by kubelet
+basically as before, based on some combination of explicitly-provided
+IPs, DNS, and IPs from the default interface, except that now it is
+based on `--node-ips` rather than `--node-ip`, and it will potentially
+add two `Addresses` rather than just one.
+
+Once `Node.Status.Addresses` is set, `Node.Status.NodeIPs` is then set
+from that and `--node-ips` as follows:
+
+  - Any explicitly-specified IP address in `--node-ips` must be
+    present in `Node.Status.Addresses`, or else kubelet will exit with
+    an error. (Note that this check is only relevant on cloud
+    platforms, since for bare metal, any value passed as `--node-ips`
+    will have just been copied to `Node.Status.Addresses`.)
+
+  - If `--node-ips` contains `ipv4` or `ipv4?`, then kubelet/ccm will
+    determine a "default IPv4 address" by finding the first IPv4
+    `InternalIP` address, or else the first IPv4 `ExternalIP` address
+    if there are no IPv4 `InternalIP` addresses. If there is no
+    default IPv4 address, then that element is removed from the list
+    in the case of `ipv4?`, or kubelet exits with an error in the case
+    of `ipv4`.
+
+  - Likewise, if `--node-ips` contains `ipv6` or `ipv6?`, then
+    kubelet/ccm will determine a "default IPv6 address" in the same
+    way, and replace the element, remove it, or error out.
+
+  - If the list is now empty (ie, it previously contained only
+    optional elements and we weren't able to default any of them) then
+    kubelet exits with an error. Otherwise, it sets
+    `Node.Status.NodeIPs` with the value.
+
+FIXME: the "kubelet exits with an error" text above assumes kubelet is
+the one setting the field. When cloud-controller-manager is setting it
+we probably need to have it just set the field to the bad value, and
+then have kubelet look at it, see that it doesn't match `--node-ips`,
+and then exit. Alternatively, maybe we should just mark the node Not
+Ready rather than having kubelet exit?
+
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy.  Anything
+that would count as tricky in the implementation and anything particularly
+challenging to test should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations).  Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+This is impossible to do full e2e testing of since that would require
```

Comment:
I would say that each cloud-provider should implement their e2e test.
We can do e2e testing for no cloud-controller-manager environments with KIND